File: clients/macos/vellum-assistant/Features/Chat/MarkdownFormatting.swift
#if os(macOS)
import AppKit

/// Pure helpers for applying markdown formatting markers around a text
/// selection. Used by ``ComposerTextView`` to handle Cmd+B / Cmd+I /
/// Cmd+Shift+X / Cmd+Shift+C shortcuts.
///
/// The logic mirrors the web implementation in
/// `vellum-assistant-platform/web/src/components/app/assistant/ChatComposer/markdown-formatting.ts`
/// so both platforms produce identical results for the same input.
enum MarkdownFormatting {

    struct Result: Equatable {
        let text: String
        let selectionStart: Int
        let selectionEnd: Int
    }

    /// Apply a markdown marker around the selected range in `text`.
    ///
    /// - **Selection present**: wraps with markers (`**selected**`), or
    ///   removes them if the selection is already wrapped (toggle-off).
    /// - **Empty cursor**: inserts paired markers with cursor between
    ///   (`**|**`).
    ///
    /// All indices are UTF-16 offsets to match `NSTextView.selectedRange()`.
    static func apply(
        text: String,
        selectionStart: Int,
        selectionEnd: Int,
        marker: String
    ) -> Result {
        let utf16 = Array(text.utf16)
        let markerUTF16 = Array(marker.utf16)
        let mLen = markerUTF16.count

        if selectionStart == selectionEnd {
            // Empty cursor — insert paired markers with cursor between.
            var result = utf16
            result.insert(contentsOf: markerUTF16, at: selectionStart)
            result.insert(contentsOf: markerUTF16, at: selectionStart + mLen)
            let newText = String(utf16CodeUnits: result, count: result.count)
            let cursor = selectionStart + mLen
            return Result(text: newText, selectionStart: cursor, selectionEnd: cursor)
        }

        // Check if the selection is already wrapped by this marker.
        let beforeStart = selectionStart - mLen
        let afterEnd = selectionEnd + mLen
        if beforeStart >= 0,
           afterEnd <= utf16.count,
           Array(utf16[beforeStart..<selectionStart]) == markerUTF16,
           Array(utf16[selectionEnd..<afterEnd]) == markerUTF16 {
            // Toggle off — remove the wrapping markers.
            var result = utf16
            result.removeSubrange(selectionEnd..<afterEnd)
            result.removeSubrange(beforeStart..<selectionStart)
            let newText = String(utf16CodeUnits: result, count: result.count)
            return Result(
                text: newText,
                selectionStart: beforeStart,
                selectionEnd: beforeStart + (selectionEnd - selectionStart)
            )
        }

        // Wrap selection with markers.
        var result = utf16
        result.insert(contentsOf: markerUTF16, at: selectionEnd)
        result.insert(contentsOf: markerUTF16, at: selectionStart)
        let newText = String(utf16CodeUnits: result, count: result.count)
        return Result(
            text: newText,
            selectionStart: selectionStart + mLen,
            selectionEnd: selectionEnd + mLen
        )
    }

    /// Match a key event to a markdown formatting marker, or return `nil`
    /// if the event doesn't correspond to a formatting shortcut.
    ///
    /// - Cmd+B → `**` (bold)
    /// - Cmd+I → `*` (italic)
    /// - Cmd+Shift+X → `~~` (strikethrough)
    /// - Cmd+Shift+C → `` ` `` (inline code)
    static func matchShortcut(modifiers: NSEvent.ModifierFlags, key: String) -> String? {
        let mods = modifiers.intersection([.shift, .command, .control, .option])
        let lowered = key.lowercased()

        if mods == .command {
            switch lowered {
            case "b": return "**"
            case "i": return "*"
            default: return nil
            }
        }

        if mods == [.command, .shift] {
            switch lowered {
            case "x": return "~~"
            case "c": return "`"
            default: return nil
            }
        }

        return nil
    }
}
#endif

File: apps/web/src/domains/chat/components/chat-composer/markdown-formatting.ts
/**
 * Pure helpers for wrapping textarea selections with markdown syntax markers.
 * Exported for unit tests — the ChatComposer `onKeyDown` handler delegates to
 * {@link applyMarkdownFormatting} to keep behavior testable without a DOM harness.
 */

interface FormattingResult {
  text: string;
  selectionStart: number;
  selectionEnd: number;
}

/**
 * Wrap or unwrap the current textarea selection with a markdown syntax marker.
 *
 * Behavior:
 * - **Selected text already wrapped**: removes the markers (toggle off).
 * - **Selected text not wrapped**: wraps with markers (toggle on).
 * - **No selection (collapsed cursor)**: inserts paired markers with cursor between them.
 */
export function applyMarkdownFormatting(
  text: string,
  selectionStart: number,
  selectionEnd: number,
  marker: string,
): FormattingResult {
  const len = marker.length;
  const selected = text.slice(selectionStart, selectionEnd);

  if (selectionStart === selectionEnd) {
    // No selection — insert paired markers with cursor between them
    const before = text.slice(0, selectionStart);
    const after = text.slice(selectionEnd);
    return {
      text: before + marker + marker + after,
      selectionStart: selectionStart + len,
      selectionEnd: selectionStart + len,
    };
  }

  // Check if selection is already wrapped with the marker
  const beforeStart = selectionStart - len;
  const afterEnd = selectionEnd + len;
  if (
    beforeStart >= 0 &&
    afterEnd <= text.length &&
    text.slice(beforeStart, selectionStart) === marker &&
    text.slice(selectionEnd, afterEnd) === marker
  ) {
    // Toggle off — remove markers
    const before = text.slice(0, beforeStart);
    const after = text.slice(afterEnd);
    return {
      text: before + selected + after,
      selectionStart: beforeStart,
      selectionEnd: beforeStart + selected.length,
    };
  }

  // Wrap selection with markers
  const before = text.slice(0, selectionStart);
  const after = text.slice(selectionEnd);
  return {
    text: before + marker + selected + marker + after,
    selectionStart: selectionStart + len,
    selectionEnd: selectionEnd + len,
  };
}

interface FormattingShortcut {
  marker: string;
  key: string;
  shiftKey: boolean;
}

/**
 * Map of formatting keyboard shortcuts following Slack/Discord/GitHub conventions:
 * - Ctrl/Cmd+B → Bold (`**`)
 * - Ctrl/Cmd+I → Italic (`*`)
 * - Ctrl/Cmd+Shift+X → Strikethrough (`~~`)
 * - Ctrl/Cmd+Shift+C → Inline code (`` ` ``)
 */
export const FORMATTING_SHORTCUTS: FormattingShortcut[] = [
  { marker: "**", key: "b", shiftKey: false },
  { marker: "*", key: "i", shiftKey: false },
  { marker: "~~", key: "x", shiftKey: true },
  { marker: "`", key: "c", shiftKey: true },
];

/**
 * Check if a keyboard event matches a formatting shortcut and return the marker.
 * Returns `null` if the event does not match any formatting shortcut.
 */
export function matchFormattingShortcut(event: {
  key: string;
  metaKey: boolean;
  ctrlKey: boolean;
  shiftKey: boolean;
}): string | null {
  const hasModifier = event.metaKey || event.ctrlKey;
  if (!hasModifier) return null;

  const keyLower = event.key.toLowerCase();
  for (const shortcut of FORMATTING_SHORTCUTS) {
    if (keyLower === shortcut.key && event.shiftKey === shortcut.shiftKey) {
      return shortcut.marker;
    }
  }
  return null;
}
